openwisp_firmware_upgrader/static/firmware-upgrader/js/batch-upgrade-progress.js

```diff
@@ -0,0 +1,433 @@
+      requestCurrentBatchState(batchUpgradeProgressWebSocket);
+    }
+  });
+
+  batchUpgradeProgressWebSocket.addEventListener("close", function (e) {
+    batchUpgradeOperationsInitialized = false;
+    if (e.code === 1006) {
+      console.error("WebSocket closed");
+    }
+  });
+
+  batchUpgradeProgressWebSocket.addEventListener("error", function (e) {
+    console.error("WebSocket error occurred", e);
+  });
+
+  batchUpgradeProgressWebSocket.addEventListener("message", function (e) {
+    try {
+      let data = JSON.parse(e.data);
+      if (data.type === "batch_status") {
+        updateBatchProgress(data);
+      } else if (data.type === "operation_progress") {
+        updateBatchOperationProgress(data);
+      } else if (data.type === "operation_update") {
+        updateBatchOperationProgress({
+          operation_id: data.operation.id,
+          status: data.operation.status,
+          progress: data.operation.progress,
+          modified: data.operation.modified,
+        });
+      }
+    } catch (error) {
+      console.error("Error parsing WebSocket message:", error);
+    }
+  });
+  batchUpgradeProgressWebSocket.open();
+}
+
+function updateBatchProgress(data) {
+  let $ = django.jQuery;
+  let mainProgressElement = $(".batch-main-progress");
+  if (mainProgressElement.length > 0) {
+    let progressPercentage =
+      data.total > 0 ? Math.round((data.completed / data.total) * 100) : 0;
+    let statusClass = (data.status || "").replace(/\s+/g, "-");
+    let showPercentageText = true;
+
+    if (data.status === "success") {
+      progressPercentage = 100;
+      statusClass = "completed-successfully";
+      showPercentageText = true;
+    } else if (data.status === "cancelled") {
+      progressPercentage = 100;
+      statusClass = "cancelled";
+      showPercentageText = false;
+    } else if (data.status === "failed") {
+      let successfulOpsCount = $("#result_list tbody tr").filter(function () {
+        let statusText = $(this).find(".status-cell .status-content").text().trim();
+        return statusText === "success" || statusText === "completed successfully";
+      }).length;
+
+      // Also check individual operation containers for success
+      if (successfulOpsCount === 0) {
+        $("#result_list tbody tr").each(function () {
+          let statusContainer = $(this).find(".upgrade-status-container");
+          if (
+            statusContainer.length &&
+            statusContainer.find(".upgrade-progress-fill.success").length
+          ) {
+            successfulOpsCount++;
+          }
+        });
+      }
+      if (successfulOpsCount > 0) {
+        // Some operations succeeded - partial success (orange)
+        progressPercentage = 100;
+        statusClass = "partial-success";
+        showPercentageText = false;
+      } else {
+        // All operations failed - total failure (red)
+        progressPercentage = 100;
+        statusClass = "failed";
+        showPercentageText = false;
```

Comment:
(same here) I see this pattern of string comparisons frequently in our `.js` files. Can we make this cleaner by using a `const object` (since JS doesn't have native enums)?